File: javascript/165.compare-version-numbers.js
/*
 * @lc app=leetcode id=165 lang=javascript
 *
 * [165] Compare Version Numbers
 */

// @lc code=start
/**
 * @param {string} version1
 * @param {string} version2
 * @return {number}
 */
var compareVersion = function(version1, version2) {
  // tediously compare...
  // O(n+m) time
  // O(n+m) space
  let v1Nums = version1.split(".").map(v => parseInt(v))
  let v2Nums = version2.split(".").map(v => parseInt(v))

  let range = Math.max(v1Nums.length, v2Nums.length)
  for (let i = 0; i < range; i++) {
    let v1Num = v1Nums[i] ?? 0
    let v2Num = v2Nums[i] ?? 0

    if (v1Num > v2Num) return 1
    if (v1Num < v2Num) return -1
  }

  return 0
};
// @lc code=end

File: cpp/165.compare-version-numbers.cpp
/*
 * @lc app=leetcode id=165 lang=cpp
 *
 * [165] Compare Version Numbers
 */

// @lc code=start
class Solution
{
public:
    int compareVersion(string version1, string version2)
    {
        // iterate over both versions from every '.'
        // build substrings to compare against
        // O(n) time O(n) space
        int indexOne = 0;
        int indexTwo = 0;
        while (indexOne < version1.size() || indexTwo < version2.size())
        {
            string sOne = "", sTwo = "";
            // build substring one for curr most significant version number
            while (indexOne < version1.size() && version1[indexOne] != '.')
            {
                if (sOne.size() == 0 && version1[indexOne] == '0')
                {
                    indexOne++;
                    continue;
                }
                sOne += version1[indexOne];
                indexOne++;
            }
            // build substring one for curr most significant version number
            while (indexTwo < version2.size() && version2[indexTwo] != '.')
            {
                if (sTwo.size() == 0 && version2[indexTwo] == '0')
                {
                    indexTwo++;
                    continue;
                }
                sTwo += version2[indexTwo];
                indexTwo++;
            }

            // compare substrings
            int vNumOne = atoi(sOne.c_str());
            int vNumTwo = atoi(sTwo.c_str());
            if (vNumOne > vNumTwo)
                return 1;
            if (vNumTwo > vNumOne)
                return -1;

            indexOne++;
            indexTwo++;
        }
        return 0;
    }
};
// @lc code=end
